### DISCUSSION TITLE
async await Nhibernate 5.3.3

### DISCUSSION THREAD

**Discussion by HakanKaraoglu** [nhibernate/nhibernate-core#2588] [General] [answered]
I am using Nhibernate 5.3.3 version. I have two different loops. I get "object referance error" in parallel loop. My queries are running in the normal loop. I do not understand what the logic could be.

Sample code with errors :

 ```
Parallel.For(0, 100, async act =>
            {
                var ws = await _workspaceRepository.GetWorkspaceByIdAsync(66);
                System.Console.WriteLine($"GetWorkspaceByIdAsync : {ws.Name}");
                var keywords = await _workspaceKeywordRepository.GetForbiddenKeywordsByWorkspaceAsync(66);
                System.Console.WriteLine($"GetForbiddenKeywordsByWorkspaceAsync : {keywords.Count}");
                var IsforbiddenKeyword = await HasForbiddenKeywordAsync(66, "facebook", "instagram");
                System.Console.WriteLine($"HasForbiddenKeywordAsync : {IsforbiddenKeyword}");
            });
```

Successful running code sample :

```
for (int i = 0; i < 100; i++)
            {
                var ws = await _workspaceRepository.GetWorkspaceByIdAsync(66);
                System.Console.WriteLine($"GetWorkspaceByIdAsync : {ws.Name}");
                var keywords = await _workspaceKeywordRepository.GetForbiddenKeywordsByWorkspaceAsync(66);
                System.Console.WriteLine($"GetForbiddenKeywordsByWorkspaceAsync : {keywords.Count}");
                var IsforbiddenKeyword = await HasForbiddenKeywordAsync(66, "facebook", "instagram");
                System.Console.WriteLine($"HasForbiddenKeywordAsync : {IsforbiddenKeyword}");
            }
```

**Comment by fredericDelaporte** [comment, marked as answer]
The session is not thread safe, do not use a session shared for parallel processing. This is unsupported and will likely never be supported.

Any async call involving a session must be awaited prior to doing anything else on the session.

See [Threads and connection](https://nhibernate.info/doc/nhibernate-reference/transactions.html#transactions-threads):
> Since NHibernate 5.0, the session and its queries IO bound methods have async counterparts. Each call to an async method must be awaited before further interacting with the session or its queries.
